File: app/src/main/java/com/mss/imagesearcher/view/imagelist/ImageListAdapter.java
package com.mss.imagesearcher.view.imagelist;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.mss.imagesearcher.App;
import com.mss.imagesearcher.R;
import com.mss.imagesearcher.presenter.main.IRvImageListPresenter;
import com.mss.imagesearcher.view.helpers.ImageLoader;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ImageListAdapter extends RecyclerView.Adapter<ImageListAdapter.ViewHolder> {

    private final IRvImageListPresenter rvPresenter;
    private OnItemClickListener onItemClickListener;

    @Inject
    ImageLoader imageLoader;

    public ImageListAdapter(@NonNull final IRvImageListPresenter rvPresenter) {
        App.getAppComponent().inject(this);
        this.rvPresenter = rvPresenter;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new ViewHolder(LayoutInflater
                .from(viewGroup.getContext())
                .inflate(R.layout.imagelist_list_item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int position) {
        rvPresenter.bindView(viewHolder);
    }

    @Override
    public int getItemCount() {
        return rvPresenter.getItemCount();
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }


    public interface OnItemClickListener {
        void onItemClick(@NonNull View view, int position);
    }


    public class ViewHolder extends RecyclerView.ViewHolder implements IImageListViewHolder {

        @BindView(R.id.pbItem)
        public ProgressBar pbItem;

        @BindView(R.id.ivItem)
        public ImageView ivItem;

        ViewHolder(@NonNull final View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @OnClick(R.id.ivItem)
        public void onImageViewClick(View view) {
            if (onItemClickListener != null) {
                onItemClickListener.onItemClick(view, getAdapterPosition());
            }
        }

        @Override
        public int getPos() {
            return getAdapterPosition();
        }

        @Override
        public void setImage(String imageURL) {
            imageLoader.load(itemView.getContext(),
                    imageURL,
                    ivItem,
                    () -> rvPresenter.onImageLoaded(ViewHolder.this),
                    () -> rvPresenter.onImageLoadFailed(ViewHolder.this));
        }

        @Override
        public void showProgress(boolean visible) {
            if (visible) {
                pbItem.setVisibility(View.VISIBLE);
            } else {
                pbItem.setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/com/mss/imagesearcher/view/imagelist/ImageListAdapter.kt
package com.mss.imagesearcher.view.imagelist

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.mss.imagesearcher.App
import com.mss.imagesearcher.R
import com.mss.imagesearcher.presenter.imagelist.IRvImageListPresenter
import com.mss.imagesearcher.view.helpers.ImageLoader
import kotlinx.android.synthetic.main.imagelist_list_item.view.*
import javax.inject.Inject

class ImageListAdapter(private val rvPresenter: IRvImageListPresenter) : RecyclerView.Adapter<ImageListAdapter.ViewHolder>() {

    private var onItemClickListener: OnItemClickListener? = null

    @Inject
    lateinit var imageLoader: ImageLoader

    init {
        App.appComponent.inject(this)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        return ViewHolder(LayoutInflater
                .from(viewGroup.context)
                .inflate(R.layout.imagelist_list_item, viewGroup, false))
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        rvPresenter.bindView(viewHolder)
    }

    override fun getItemCount(): Int {
        return rvPresenter.itemCount
    }

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener) {
        this.onItemClickListener = onItemClickListener
    }


    interface OnItemClickListener {
        fun onItemClick(view: View, position: Int)
    }


    inner class ViewHolder constructor(itemView: View) : RecyclerView.ViewHolder(itemView), IImageListViewHolder {

        override val pos: Int
            get() = adapterPosition

        init {
            itemView.ivItem.setOnClickListener { onItemClickListener?.onItemClick(it, adapterPosition) }
        }

        override fun setImage(imageURL: String) {
            imageLoader.load(itemView.context,
                    imageURL,
                    itemView.ivItem!!,
                    { rvPresenter.onImageLoaded(this@ViewHolder) },
                    { rvPresenter.onImageLoadFailed(this@ViewHolder) })
        }

        override fun showProgress(visible: Boolean) {
            if (visible) {
                itemView.pbItem!!.visibility = View.VISIBLE
            } else {
                itemView.pbItem!!.visibility = View.GONE
            }
        }
    }
}
